docs/proposals/petset.md

```diff
@@ -0,0 +1,533 @@
+each new identity. The controller must create supporting resources for the pod before creating the
+pod. If a supporting resource with the appropriate name already exists, the controller should treat that as
+creation succeeding. If a supporting resource cannot be created, the controller should flag an error to
+status, back-off (like a scheduler or replication controller), and try again later.
+
+When a PetSet is scaled down, the pod for the removed indentity should be deleted. It is less clear what the
+controller should do to supporting resources. If every pod requires a PV, and a user accidentally scales
+up to N=200 and then back down to N=3, leaving 197 PVs lying around may be undesirable (potential for
+abuse). On the other hand, a cluster of 5 that is accidentally scaled down to 3 might irreparably destroy
+the cluster if the PV for identities 4 and 5 are deleted (may not be recoverable). For the initial proposal,
+leaving the supporting resources is the safest path (safety first) with a potential future policy applied
+to the PetSet for how to manage supporting resources (DeleteImmediately, GarbageCollect, Preserve).
+
+OPEN: Should information about the state of the pods in the PetSet be reflected in PetSet status? Summarizing
+  which pods match the current PetSet state would be extremely valuable.
+
+
+### Parameterizing pod templates and supporting resources
+
+Since each pod needs a unique and distinct identity, and the pod needs to know its own identity, the
+PetSet must allow a pod template to be parameterized by the identity assigned to the pod. The pods that
+are created should be easily identified by their cluster membership.
+
+
+```
+type PetSetSpec struct {
+    // Number of instances in the set
+    Replicas *int
+
+    // The template to use for the pod in the set
+    Template *PodTemplateSpec
+}
+
+// Proposal 1a: Use the name of the pod as the identity, use downward API for getting identity
+// Downside: all fields that want to leverage the identity need equivalent valueFrom semantics
+{
+    "spec": {
+        "template": {
+            "metadata": {"generateName": "pod-name"}, // assume identity is automatically added to pod name
+            "spec": {
+                "containers": [{
+                    "env": [{
+                        "name": "IDENTITY",
+                        "valueFrom": {"fieldRef": {"fieldPath": "metadata.name"}
+                        }
+                    }]
+                }]
+            }
+        }
+    }
+}
+
+// Proposal 1b: Add an annotation to each pod when the controller creates it
+// Downside: all fields that want to leverage the identity need equivalent valueFrom semantics
+{
+    "spec": {
+        "template": {
+            "spec": {
+                "containers": [{
+                    "env": [{
+                        "name": "IDENTITY",
+                        "valueFrom": {"fieldRef": {"fieldPath": "metadata.annotations[alpha.kubernetes.io/identity]"}}
```

Comment:
If I only get one annotation different per pod, then I cannot easily do:
`./command -a $A -b $B`.

And, if I only get a base 0 or base 1 index, then I need another step to lookup a string from an integer.

In other words, I cannot tell if I like this plan or not based on how it is described.  It needs a `.spec.parameters` section.